va/validation-authority.go

```diff
@@ -498,20 +498,28 @@ func (va *ValidationAuthorityImpl) validateDNS01(ctx context.Context, identifier
 }
 
 func (va *ValidationAuthorityImpl) checkCAA(ctx context.Context, identifier core.AcmeIdentifier) *probs.ProblemDetails {
-	// Check CAA records for the requested identifier
-	present, valid, err := va.checkCAARecords(ctx, identifier)
-	if err != nil {
-		return bdns.ProblemDetailsFromDNSError(err)
-	}
-	// AUDIT[ Certificate Requests ] 11917fa4-10ef-4e0d-9105-bacbe7836a3c
-	va.log.AuditInfo(fmt.Sprintf("Checked CAA records for %s, [Present: %t, Valid for issuance: %t]", identifier.Value, present, valid))
-	if !valid {
-		return &probs.ProblemDetails{
-			Type:   probs.ConnectionProblem,
-			Detail: fmt.Sprintf("CAA record for %s prevents issuance", identifier.Value),
+	var prob *probs.ProblemDetails
+	if va.caaClient != nil {
+		prob = va.checkCAAService(ctx, identifier)
+	} else {
+		present, valid, err := va.checkCAARecords(ctx, identifier)
+		if err != nil {
+			prob = bdns.ProblemDetailsFromDNSError(err)
+		} else {
+			// AUDIT[ Certificate Requests ] 11917fa4-10ef-4e0d-9105-bacbe7836a3c
+			va.log.AuditInfo(fmt.Sprintf("Checked CAA records for %s, [Present: %t, Valid for issuance: %t]", identifier.Value, present, valid))
+			if !valid {
+				prob = &probs.ProblemDetails{
+					Type:   probs.CAAProblem,
+					Detail: fmt.Sprintf("CAA record for %s prevents issuance", identifier.Value),
+				}
+			}
 		}
 	}
-	return nil
+	if va.caaDR != nil && prob != nil && prob.Type == probs.ConnectionProblem {
```

Comment:
This seems like it is using ProblemDetails as internal error handling. We should avoid doing that. ProblemDetails are intended for an error that we know we can pass all the way back to the user without modification. If there's an internal error we may attempt to handle, we should use another error type, and handle it internally. We can then handle the results at a higher level and turn them into ProblemDetails if necessary.